### PR TITLE
filters: add bridge support for data invocations

### PR DESCRIPTION
**Pull request by goaway**
Description: Add initial support for HTTP body data chunks to be bridged to platform filter invocations.
Risk Level: Moderate
Testing: Local

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

File: library/common/extensions/filters/http/platform_bridge/c_types.cc [hunk 1]
```diff
@@ -13,3 +13,10 @@ const envoy_filter_headers_status_t kEnvoyFilterHeadersStatusContinueAndEndStrea
 const envoy_filter_headers_status_t kEnvoyFilterHeadersStatusStopAllIterationAndBuffer =
     static_cast<envoy_filter_headers_status_t>(
         Envoy::Http::FilterHeadersStatus::StopAllIterationAndBuffer);
+
+const envoy_filter_data_status_t kEnvoyFilterDataStatusContinue =
```
Comment (junr03) [inline library/common/extensions/filters/http/platform_bridge/c_types.cc:17]: nit: is the `k` prefix a common c-style convention?

File: library/common/extensions/filters/http/platform_bridge/c_types.h [hunk 2]
```diff
@@ -24,11 +24,10 @@ typedef struct {
 /**
  * Return codes for on-data filter invocations. @see envoy/http/filter.h
  */
-typedef enum {
-  ENVOY_FILTER_DATA_STATUS_CONTINUE = 0,
-  ENVOY_FILTER_DATA_STATUS_STOP_ITERATION_AND_BUFFER,
-  ENVOY_FILTER_DATA_STATUS_STOP_ITERATION_NO_BUFFER,
-} envoy_filter_data_status_t;
+typedef int envoy_filter_data_status_t;
```
Comment (junr03) [inline library/common/extensions/filters/http/platform_bridge/c_types.h:27]: Can we document this style decision in STYLE.md?

File: library/common/extensions/filters/http/platform_bridge/filter.cc [hunk 3]
```diff
@@ -47,15 +48,37 @@ Http::FilterHeadersStatus PlatformBridgeFilter::onHeaders(Http::HeaderMap& heade
...
+  envoy_data in_data = Buffer::Utility::toBridgeData(data);
+  envoy_filter_data_status result = on_data(in_data, end_stream, platform_filter_->context);
+  Http::FilterDataStatus status = static_cast<Http::FilterDataStatus>(result.status);
+  // Current platform implementations expose immutable data, thus any modification necessitates a
+  // full copy. If the returned buffer is identical, we assume no modification was made and elide
+  // the copy here. See also https://github.com/lyft/envoy-mobile/issues/949 for potential future
+  // optimization.
+  if (in_data.bytes != result.data.bytes) {
+    data.drain(data.length());
+    data.addBufferFragment(*Buffer::BridgeFragment::createBridgeFragment(result.data));
```
Comment (junr03) [inline library/common/extensions/filters/http/platform_bridge/filter.cc:67]: I still think that this wrapping class is such a nifty approach, love it.

File: library/common/extensions/filters/http/platform_bridge/c_types.cc (hunk 1 above)
Comment (goaway) [reply library/common/extensions/filters/http/platform_bridge/c_types.cc:17]: It is _a_ convention that has been used in C, C++, Objective-C and Swift. It's the nearest I could come to a style shared by and recognizable across all four languages.

See, for example:
https://google.github.io/styleguide/cppguide.html#Variable_Comments

Note, the Google Style Guide for Objective-C, _no longer_ recommends prefixing global constants with `k` (where it was once quite prevalent), but again I was going for something that would be easily recognizable across all the languages/environments.

File: library/common/extensions/filters/http/platform_bridge/c_types.h (hunk 2 above)
Comment (goaway) [reply library/common/extensions/filters/http/platform_bridge/c_types.h:27]: Sure, but just to clarify, which do you mean?

Comment (junr03) [reply library/common/extensions/filters/http/platform_bridge/c_types.h:27]: The new approach with the static cast

Comment (goaway) [reply library/common/extensions/filters/http/platform_bridge/c_types.h:27]: Oh, sure.

Comment (goaway) [reply library/common/extensions/filters/http/platform_bridge/c_types.h:27]: Maybe as part of this PR?
https://github.com/lyft/envoy-mobile/pull/966

Comment (junr03) [reply library/common/extensions/filters/http/platform_bridge/c_types.h:27]: will document in https://github.com/lyft/envoy-mobile/pull/966/files

**Pull request merged**